### PR TITLE
netapi: adapt to pkt_t

### PR DESCRIPTION
**Pull request by authmillenon**
Simplifies netapi quiet a bit.

Depends on ~~#2158~~ (merged). There is also a small dependency on ~~#1968~~ (merged) or vice-versa, but I'll just will fix the PR that get's merged later ;-)

Depends on #2285

### CODE REVIEW

File: sys/include/pkt.h [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+#include <inttypes.h>
+
+#include "clist.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Definition of protocol families to determine the type of the packet or which protocols a network device (see @ref netdev) or protocol layer (see @ref netapi) can handle
```
Comment (haukepetersen) [inline sys/include/pkt.h]: line length...

File: sys/include/pkt.h [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    PKT_PROTO_TCP           = 0x0006,   /**< TCP. */
+    PKT_PROTO_CCNL          = 0x0007,   /**< CCN lite. */
+
+    /**
+     * @brief   CC110x frame format protocol
+     *
+     * @details Sends frames as defined by cc110x_packet_t.
+     */
+    PKT_PROTO_CC110X        = 0x0008,
+} pkt_proto_t;
```
Comment (haukepetersen) [inline sys/include/pkt.h]: How about we include some (upcoming) protocols from the beginning, i.e. IPv4, Ethernet, BLE?

File: sys/include/pkt.h [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+ *          unspecified type to work with @ref clist.h.
+ *
+ * @note    This type implements its own list implementation because of the way
+ *          it is stored in the packet buffer.
+ */
+typedef struct __attribute__((packed)) pkt_hlist_t {    /* packed to be aligned
+                                                         * correctly in static
+                                                         * packet buffer */
+    pktsize_t header_len;       /**< the length of the header in byte. */
+    void *header_data;          /**< the data of the header */
```
Comment (haukepetersen) [inline sys/include/pkt.h]: very dangerous: this could lead to hard_faults on Cortex-M0 platforms, as the `*header_data` pointer is not aligned on a 32-bit border, or do I have something wrong here?

**Comment by haukepetersen** [comment]
I like the idea of the pkt type. I would actually strongly opt for removing the (redundant) *_hlist_* data-structue and also remove the hlist from the functions naming -> keep it as simple as possible!

File: sys/include/pkt.h (hunk 2 above)
Comment (authmillenon) [inline sys/include/pkt.h]: Good idea, will do.

File: sys/include/pkt.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/pkt.h]: The packed attribute should prevent that or am I wrong?

**Comment by authmillenon** [comment]
But headers are no packets. As someone not familiar with this I would be very confused if I see packets get stuffed into packets. Furthermore, the semantics become very different in the packet buffer (only `pkt_t` can created, `pkt_hlist_t`s can only be prepended/removed from `pkt_t`.

Think of the `pkt_t` as head type of the of a list with `pkt_hlist_t` nodes and `pktbuf` as its factory.

**Comment by authmillenon** [comment]
TL;DR while removing `pkt_list_t` would reduce complexity in terms of LOC, I would find the API less understandable.

**Comment by authmillenon** [comment]
Afterthought: You commented on the wrong PR. Please discuss `pkt_t` in #2158 ;-)

File: sys/include/pkt.h (hunk 3 above)
Comment (authmillenon) [inline sys/include/pkt.h]: But you're right: better save than sorry

**Comment by authmillenon** [comment]
Rebased to make changes apparent here, too. ;-)

**Comment by haukepetersen** [comment]
I thought about this a bit more: I would actually vote for renaming the thing completely to something like `chunk_t`, getting away from packets/headers all together. So all you do in the network stack is allocating chunks of memory in the `chunkbuf`. Or maybe we can call it `crumb_t / crumbstore` or something...

Anyway, as I see it the buffer can be used for more than just packet/header buffering, why not use it also to store addresses or whatever?

Regarding the API: I don't think we need a dedicated API for headers, by just working on a generic data structure we simplify/shorten code with the same results - in the end it's not the genuine user that is working inside the network stack - but people who know at least to some extend what they are doing...

**Comment by haukepetersen** [comment]
oh, just noticed: this comment should have probably been posted for #2158, sorry

**Comment by LudwigOrtmann** [comment]
@authmillenon please rebase

**Comment by authmillenon** [comment]
Rebased to #2285 and added it as dependency in description + ported `nomac` to the api change.

**Comment by authmillenon** [comment]
Closed in favor for #2400
